### PR TITLE
Avoid segfault caused by failing to swap back materials

### PR DESCRIPTION
**Pull request by jwpeterson**
This should fix the segfault issue discovered by @WilkAndy in #7991.

### CODE REVIEW

File: framework/include/utils/SwapBackSentinel.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ */
+class SwapBackSentinel
+{
+public:
+  /**
+   * SwapBackFunction is a typedef for a pointer to an FEProblem
+   * member function taking a THREAD_ID and returning void.  All the
+   * FEProblem::swapBackMaterialXXX() members have this signature.
+   */
+  typedef void (FEProblem::*SwapBackFunction)(THREAD_ID);
```
Comment (permcody) [inline framework/include/utils/SwapBackSentinel.h]: Fine, but prefer c++11 syntax.

**Review by permcody — APPROVED** [review]
This looks fine to me and is definitely an improvement over what we have now. I'll see if @friedmud has any comments and we can merge this early next week in either case.

File: framework/include/utils/SwapBackSentinel.h (hunk 1 above)
Comment (jwpeterson) [inline framework/include/utils/SwapBackSentinel.h]: I looked into it, and the syntax was [way worse](https://github.com/jwpeterson/moose/commit/1a5b692bc839c114e5d476eb7b9f5767dc20ebdd) for dealing with member function pointers IMO.  I'd be happy to change it if you can figure out an interface that is cleaner than the C++03 way for this simple application...

Comment (jwpeterson) [inline framework/include/utils/SwapBackSentinel.h]: Oh, I see.  Yeah that would probably be better...

File: framework/include/utils/SwapBackSentinel.h [hunk 2]
```diff
@@ -40,7 +40,7 @@ class SwapBackSentinel
    * member function taking a THREAD_ID and returning void.  All the
    * FEProblem::swapBackMaterialXXX() members have this signature.
    */
-  typedef void (FEProblem::*SwapBackFunction)(THREAD_ID);
+  using SwapBackFunction = void (FEProblem::*)(THREAD_ID);
```
Comment (permcody) [inline framework/include/utils/SwapBackSentinel.h]: I think that's more visually appealing - Thanks!

**Comment by friedmud** [comment]
I like the idea (and we may need more of this in other places too... do we restore the Auxiliary solution if we're half-way through computing AuxKernels and there is an exception?).

However: did you happen to do any timing to make sure that the creation/destruction of these objects isn't showing up?  At least a quick run through Instruments just to make sure we don't see these guys.

I wouldn't expect them to show up... they are small and should be completely on the stack... but I've been surprised before....

**Comment by jwpeterson** [comment]
I'll still verify in Instruments whether the new objects show up at all...

**Comment by jwpeterson** [comment]
I profiled simple_diffusion on a 500x500 mesh.  This problem doesn't actually have Materials, so I figured it would be good for just showing the overhead introduced by the object creation/destruction.  The attached screenshot drills down into ComputeResidualThread::onElement(), where one of the new `Sentinel` objects is created.  I think the constructor/destructor calls may be getting inlined (?) since they doesn't explicitly show up in the trace.  Let me know if there is some other case it would be useful to profile...

simple_diffusion_500x500

**Comment by jwpeterson** [comment]
For reference, here's the profiling for `ComputeResidualThread::onElement()` from before the `Sentinel` objects were merged.  In this case it looks like the call to `_fe_problem.swapBackMaterials(_tid);` itself might be inlined, as it does not show up in the trace.  Either way, I think the main take away is that there was not much affect on the overall time required for `ComputeResidualThread::onElement()`, which for this single data point is:

Before Sentinel: 755 ms, 8.7% of total runtime.
After Sentinel: 737ms, 8.3% of total runtime.

simple_diffusion_500x500_orig

**Comment by friedmud** [comment]
Sounds good. Thanks for checking!

**Pull request merged**